File: 05/sol.py
contents = open("inp.txt").read()
# contents = open("little.txt").read()

rules, e = contents.split("\n\n")

rd = {}
for rule in rules.split():
    a, b = map(int, rule.split("|"))
    if a not in rd:
        rd[a] = []
    rd[a].append(b)

c = 0
others = []

for en in e.split():
    l = list(map(int, en.split(",")))
    t = {}
    for i, n in enumerate(l):
        t[n] = i

    fail = False

    for n, i in t.items():
        if n not in rd:
            continue
        rs = rd[n]
        for r in rs:
            if r in t and t[r] < i:
                fail = True
        if fail:
            break

    if not fail:
        c += l[len(l)//2]
    else:
        others.append(t)

print(c)

c = 0

for o in others:
    acc = []
    while len(o) > 0:
        for n, i in o.items():
            if n not in rd or not any(map(lambda k: k in o, rd[n])):
                acc.append(n)
                del o[n]
                break

    acc.reverse()
    c += acc[len(acc)//2]

print(c)

File: 05/sol.hs
import Data.Map hiding (drop, take, map, filter, delete)
import qualified Data.Map as M
import Data.Maybe
import Data.List

splitHelp :: String -> String -> String -> Maybe (String, String)
splitHelp sub ori "" = Nothing
splitHelp sub ori s
  | Just new <- stripPrefix sub s = Just (take (length ori - length s) ori, new)
  | otherwise = splitHelp sub ori $ tail s

splitSub :: String -> String -> Maybe (String, String)
splitSub sub s = splitHelp sub s s

fullSplit :: String -> String -> [String]
fullSplit sub s
  | Just (a, b) <- splitSub sub s = [a] ++ fullSplit sub b
  | otherwise = [s]

part1H :: Map Int [Int] -> [Int] -> Int
part1H m en
  | not $ any checkBounds $ assocs t = en !! (length en `div` 2)
  | otherwise = 0
    where t = fromList $ zip en [1..]
          checkAfter i v = (elem v en) && (t ! v) < i
          checkBounds (k, v) = (member k m) && (any (checkAfter v) (m ! k))

part1 :: Map Int [Int] -> [[Int]] -> Int
part1 m ent = sum $ map (part1H m) ent


part2H :: Map Int [Int] -> [Int] -> Int
part2H m en = res !! (length res `div` 2)
    where res = reverse $ helper en
          helper [] = []
          helper en = target:(helper $ delete target en)
              where Just target = (find isFree en)
                    isFree k = all (\n -> not $ elem n en) $ m ! k

part2 :: Map Int [Int] -> [[Int]] -> Int
part2 m ent = sum $ map (part2H m) $ filter ((0 ==) . (part1H m)) ent

main :: IO ()
main = do
    contents <- readFile "inp.txt"
    let [ords, entrs] = fullSplit "\n\n" contents
        m1 = fromListWith (++) $
            [(read a, [read b]) | l <- lines ords, let [a,b] = fullSplit "|" l ]
        l1 = map (\l -> map read $ fullSplit "," l) $ lines entrs
        
    print $ part1 m1 l1
    print $ part2 m1 l1

